observability/logging.go

```diff
@@ -0,0 +1,222 @@
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().Data
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_HALF_CLOSE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_HALF_CLOSE
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_TRAILER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetTrailer().Metadata)
+		grpcLogRecord.StatusCode = binlogEntry.GetTrailer().StatusCode
+		grpcLogRecord.StatusMessage = binlogEntry.GetTrailer().StatusMessage
+		grpcLogRecord.StatusDetails = binlogEntry.GetTrailer().StatusDetails
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+		callEnded = true
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CANCEL:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_CANCEL
+		callEnded = true
+	default:
+		return fmt.Errorf("unknown event type: %v", binlogEntry.Type)
+	}
+	switch binlogEntry.Logger {
+	case binlogpb.GrpcLogEntry_LOGGER_CLIENT:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_CLIENT
+	case binlogpb.GrpcLogEntry_LOGGER_SERVER:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_SERVER
+	default:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_UNKNOWN
+	}
+	if callEnded {
+		cls.removeEntry(binlogEntry.CallId)
+	}
+	// CloudLogging client doesn't return error on entry write. Entry writes
+	// don't mean the data will be uploaded immediately.
+	cls.exporter.EmitGrpcLogRecord(&grpcLogRecord)
+	return nil
+}
+
+// Close closes the cloudLoggingSink, which is a noop due to no state is
+// maintained by this object.
+func (*cloudLoggingSink) Close() error {
+	return nil
+}
+
+func newCloudLoggingSink() *cloudLoggingSink {
+	return &cloudLoggingSink{
+		callIDToUUID: make(map[uint64]string),
+	}
+}
+
+func compileBinaryLogControlString(config *configpb.ObservabilityConfig) string {
+	if config.LoggingConfig == nil {
+		return ""
+	}
+
+	var entries []string
+	for _, logFilter := range config.LoggingConfig.LogFilters {
+		// With undefined HeaderBytes or MessageBytes, the intended behavior is
+		// logging zero payload. This detail is different than binary logging.
+		entries = append(entries, fmt.Sprintf("%v{h:%v;m:%v}", logFilter.Pattern, logFilter.HeaderBytes, logFilter.MessageBytes))
+	}
+	entries = append(entries, "-google.logging.v2.LoggingServiceV2/WriteLogEntries")
+	return strings.Join(entries, ",")
+}
+
+func startLogging(config *configpb.ObservabilityConfig) {
+	if config == nil {
+		return
+	}
+	var binlogConfig = compileBinaryLogControlString(config)
+	iblog.SetLogger(iblog.NewLoggerFromConfigString(binlogConfig))
```

Comment:
Maybe this stuff should wait until `Start` is called?  Then we can return errors.  At `init` time we should just install the binary log sink in "inactive" mode.

Reply:
Renamed the function to `prepareLogging`. This won't start logging, just register the logger as we discussed.

The public `binarylog` module only provides `SetSink` and `NewTempFileSink` 2 methods. I experimented adding the controlling interface (the string) via environment variable, and use the public interface only, the result is that the env won't be picked up by `binarylog` (nothing gets logged). So... I was left between 2 choices: 1. call internal; 2. publicize `SetLogger`. I picked the former.

Reply:
I don't understand.  Why is `SetSink` not enough for our needs here?  Why do we need to set the logger?

Can we not allow the "logger" to be configured as normal, but filter the things we want to log in this module?  Instead of converting the observability config into a gRPC-internal config.